### ISSUE TITLE
New Rule: Use SELECT DISTINCT instead over grouping by all columns

### ISSUE THREAD

**Issue by jpy-git** [sqlfluff/sqlfluff#2393]
From Brooklyn Data co. SQL style guide: https://github.com/brooklyn-data/co/blob/main/sql_style_guide.md#use-select-distinct-instead-of-grouping-by-all-columns

```
Use select distinct instead of grouping by all columns.
This makes the intention clear.
```

```sql
/* Good */
select distinct
    customer_id
    , date_trunc('day', created_at) as purchase_date
from orders

/* Bad */
select
    customer_id
    , date_trunc('day', created_at) as purchase_date
from orders
group by 1, 2
```

Think this entails checking if the number of group by elements matches the number of select clause elements and if so raise then lint and apply fix.

I guess theres always the possibility the user writes some bad sql like below that will parse in sqlfluff but is actually invalid SQL, so maybe we also apply some extra sanity checks to make sure we don't try and fix these:
```sql
select
    customer_id
    , date_trunc('day', created_at) as purchase_date
from orders
group by 6, 9

-- or

select
    customer_id
    , date_trunc('day', created_at) as purchase_date
from orders
group by some_other_column, another_column
```

**Comment by tunetheweb** [comment]
Probably should just include exclude any statements with function calls in the SELECT clause as that's using the reason use GROUPING (e.g. `SELECT col1, COUNT(1) FROM table1 GROUP BY col1`).

**Comment by fmms** [comment]
It becomes more complicated than this as the projection happens last
```sql
SELECT LEFT(yearmonth, 4) AS myyear, RIGHT(yearmonth, 2) AS mymonth
FROM dbo.datedimension
GROUP BY yearmonth
```

So it is not the number of non aggregation elements in the select, rather the number of referenced columns.
